Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs

```diff
@@ -103,5 +230,51 @@ public async Task<IEnumerable<string>> ListTableNames(CancellationToken cancella
 
             return (await result.ToListAsync(cancellationToken)).Select(i => i.StringValue);
         }
+
+        private async Task<AsyncQldbSession> GetSession()
+        {
+            this.logger.LogDebug(
+                "Getting session. There are {} free sessions and {} available permits.",
+                this.sessionPool.Count,
+                this.sessionPool.BoundedCapacity - this.poolPermits.CurrentCount);
+
+            if (await this.poolPermits.WaitAsync(DefaultTimeoutInMs))
+            {
+                var session = this.sessionPool.Count > 0 ? this.sessionPool.Take() : null;
+
+                if (session == null)
+                {
+                    session = await this.StartNewSession();
+                    this.logger.LogDebug("Creating new pooled session with ID {}.", session.GetSessionId());
+                }
+
+                return session;
+            }
+            else
+            {
+                this.logger.LogError(ExceptionMessages.SessionPoolEmpty);
+                throw new QldbDriverException(ExceptionMessages.SessionPoolEmpty);
+            }
+        }
+
+        private async Task<AsyncQldbSession> StartNewSession()
```

Comment:
I'm not as familiar with .Net async as with other areas, but it seems odd to be calling another async method from an async method, doesn't that essentially start a thread for GetSession, then block that thread on another thread for StartSession which then blocks on StartSessionAsync? Can StartNewSession simply not be async?

Reply:
Async in .NET works like it does in Node.js: it propagates through the API all the way from the bottom to the top. Threads aren't necessarily started, as it's just a nice decorator for callbacks.